### PR TITLE
fix(kafka): ConsumeClaim should return when session context is done

### PR DESCRIPTION
**Pull request by zcong1993**
Signed-off-by: zcong1993 <zhangcong1992@gmail.com>

# Description

ConsumeClaim should return when `session.Context()` is done, if not, will raise `ErrRebalanceInProgress` or `read tcp <ip>:<port>: i/o timeout` when kafka rebalance. see: https://github.com/Shopify/sarama/issues/1192, and sarama updated their example https://github.com/Shopify/sarama/blob/main/examples/consumergroup/main.go#L191 .

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: internal/component/kafka/consumer.go [hunk 1]
```diff
@@ -35,26 +35,33 @@ type consumer struct {
...
-				consumer.k.logger.Infof("Successfully processed Kafka message after it previously failed: %s/%d/%d [key=%s]", message.Topic, message.Partition, message.Offset, asBase64String(message.Key))
-			}); err != nil {
-				consumer.k.logger.Errorf("Too many failed attempts at processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
-			}
-		} else {
-			err := consumer.doCallback(session, message)
-			if err != nil {
-				consumer.k.logger.Errorf("Error processing Kafka message: %s/%d/%d [key=%s]. Error: %v.", message.Topic, message.Partition, message.Offset, asBase64String(message.Key), err)
+
+	for {
```
Comment (daixiang0) [inline internal/component/kafka/consumer.go:39]: Could we have a test for this?

Comment (zcong1993) [reply internal/component/kafka/consumer.go:39]: I have tested both versions locally, both work fine, but the official code in the example has been updated, this way should be the official recommendation.

**Review by ItalyPaleAle — APPROVED** [review]

**Review by berndverst — APPROVED** [review]

File: internal/component/kafka/consumer.go [hunk 2]
```diff
@@ -35,26 +35,33 @@ type consumer struct {
...
+		select {
+		case message := <-claim.Messages():
```
[the remark this answers is not in the document]
Comment (zcong1993) [reply internal/component/kafka/consumer.go:41]: `claim.Messages()` is just go channel `<-chan *ConsumerMessage`, so we only get one message per read for sure.

[the remark this answers is not in the document]
Comment (fdejaeger) [reply internal/component/kafka/consumer.go:41]: beware that message can be nil now, if it happens that the channel is closed before the cancel happens.  The code samples are still broken.

[the remark this answers is not in the document]
Comment (zcong1993) [reply internal/component/kafka/consumer.go:41]: good catch，could you make a pr to fix this？

[the remark this answers is not in the document]
Comment (zcong1993) [reply internal/component/kafka/consumer.go:41]: > beware that message can be nil now, if it happens that the channel is closed before the cancel happens.  The code samples are still broken.

good catch，could you make a pr to fix this？

[the remark this answers is not in the document]
Comment (fdejaeger) [reply internal/component/kafka/consumer.go:41]: Sorry I have no business with dapr, I just came here from https://github.com/Shopify/sarama/issues/1192.  I'm just another sarama's consumer  and got bitten by the same problem as you.
I've filled this bug for them on the topic https://github.com/Shopify/sarama/issues/2344

**Comment by weiGaoCheng** [comment]
加上case <-session.Context().Done()后确实不再报timeout错误  但是开始报错  I0822 17:12:30.410037       8 sarama_logger.go:12] consumer/data/9 shutting down because kafka server: The requested offset is outside the range of offsets maintained by the server for the given topic/partition

**Pull request merged**